pkg/tfbridge/schema_test.go

```diff
@@ -272,6 +273,109 @@ func TestTerraformInputs(t *testing.T) {
 	}
 }
 
+func TestMakeTerraformInputMixedMaxItemsOne(t *testing.T) {
+	t.Parallel()
+	tfs := func(scalar, list bool) shim.SchemaMap {
+		stringSchema := (&schema.Schema{
+			Type: shim.TypeString,
+		}).Shim()
+		s, l := schema.Schema{}, schema.Schema{}
+		if scalar {
+			s.Type = shim.TypeList
+			s.MaxItems = 1
+			s.Elem = stringSchema
+		} else {
+			s.Type = shim.TypeString
+		}
+
+		l.Type = shim.TypeList
+		if list {
+			l.MaxItems = 1
+			l.Elem = (&schema.Schema{
+				Type: shim.TypeList,
+				Elem: stringSchema,
+			}).Shim()
+		} else {
+			l.Elem = stringSchema
+		}
+		return schema.SchemaMap{
+			"scalar": s.Shim(),
+			"list":   l.Shim(),
+		}
+	}
+	property := func(scalar, list bool) resource.PropertyMap {
+		m := resource.PropertyMap{
+			"scalar": resource.NewStringProperty("sc"),
+			"list":   resource.NewStringProperty("sc"),
+		}
+		wrap := func(key resource.PropertyKey) {
+			v := m[key]
+			arr := resource.NewArrayProperty([]resource.PropertyValue{v})
+			m[key] = arr
+		}
+		wrap("list")
+
+		if scalar {
+			wrap("scalar")
+		}
+		if list {
+			wrap("list")
+		}
+		return m
+	}
+
+	tests := map[string]struct {
+		oldsScalar, oldsList bool
+		newsScalar, newsList bool
+		expected             map[string]interface{}
+	}{
+		"scalar-adding-max-items-one": {
+			oldsScalar: false,
+			newsScalar: true,
+			expected: map[string]interface{}{
+				"list":   []interface{}{"sc"},
+				"scalar": []interface{}{"sc"},
+			},
+		},
+		"scalar-removing-max-items-one": {
+			oldsScalar: true,
+			newsScalar: false,
+			expected: map[string]interface{}{
+				"list":   []interface{}{"sc"},
+				"scalar": "sc",
+			},
+		},
+		"list-adding-max-items-one": {
+			oldsList: false,
+			newsList: true,
+			expected: map[string]interface{}{
+				"list":   []interface{}{[]interface{}{[]interface{}{"sc"}}},
```

Comment:
What's going on in this test case?

The test look great in the sense that it looks like they cover all the interesting combinations, however some attention perhaps could be given to legibility for an out-of-context engineer coming to read this code.

The property we are trying to check for and ensure is that the result of MakeTerraformInput always conforms to the Terraform schema, as violating this property leads to panics or errors in the TF layer.

It would be lovely to make it obvious in the test case. For example, here we are looking at this value (I think the value under "scalar" is not meaningfully testing things in this test case, right?):

```
[[["sc"]]]
``` 

So presumably the TF schema here is `List<List<List<string>>>`, but it's hard to see that since it's hidden behind generating functions.

[Boolean blindness](https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/) comes to mind.

Reply:
I've revised the test cases to be clearer and more readable, and added some additional tests on null handling. To avoid boolean blindness, I have removed the constructer functions.